File: src/go/ch01/1.1/l1/twos_complement.go
package main

import (
	"errors"
	"fmt"
)

// ToTwosComplement encodes x into a two's complement bitstring of given width.
// bits must be in [1, 64]. The value is truncated to the requested width.
func ToTwosComplement(x int64, bits int) string {
	if bits < 1 || bits > 64 {
		panic("bits must be between 1 and 64")
	}
	var mask uint64
	if bits == 64 {
		mask = ^uint64(0)
	} else {
		mask = (1 << bits) - 1
	}
	u := uint64(x) & mask

	// Build string high-to-low.
	out := make([]byte, bits)
	for i := bits - 1; i >= 0; i-- {
		if (u & 1) == 1 {
			out[i] = '1'
		} else {
			out[i] = '0'
		}
		u >>= 1
	}
	return string(out)
}

// FromTwosComplement decodes a two's-complement bitstring (e.g., "11111001")
// into a signed integer. The width is len(s). Returns error on invalid input.
func FromTwosComplement(s string) (int64, error) {
	if len(s) == 0 || len(s) > 64 {
		return 0, errors.New("invalid width")
	}
	var val uint64
	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '0':
			val = val << 1
		case '1':
			val = (val << 1) | 1
		default:
			return 0, fmt.Errorf("invalid character %q", s[i])
		}
	}
	bits := uint(len(s))
	top := uint64(1) << (bits - 1)
	if (val & top) == 0 {
		// positive
		return int64(val), nil
	}
	// negative: subtract 2^bits
	full := uint64(1) << bits
	return int64(val - full), nil
}

File: src/erlang/ch01/1.1/l1/src/twos_complement.erl
%% Two's complement encode/decode for fixed bit-width.
-module(twos_complement).
-export([to/2, from/2]).

%% to(X, Bits) -> "bitstring" as list of chars '0'/'1' of length Bits.
%% X :: integer(), Bits :: 1..64 (you can raise if you want more).
to(X, Bits) when is_integer(X), is_integer(Bits), Bits >= 1, Bits =< 64 ->
    Mask = (1 bsl Bits) - 1,
    U = X band Mask,                                      % truncate to width
    Bin = integer_to_list(U, 2),
    PadLen = Bits - length(Bin),
    Padding = lists:duplicate(max(PadLen, 0), $0),
    Padding ++ Bin.

%% from(Str, Bits) -> signed integer according to two's complement with given Bits.
%% Str is a binary/iodata or string of '0'/'1'; Bits = length(Str) is typical.
from(Str0, Bits) when is_integer(Bits), Bits >= 1, Bits =< 64 ->
    Str = ensure_string(Str0),
    true = (length(Str) == Bits) orelse error({bad_width, {expected, Bits}, {got, length(Str)}}),
    Value = list_to_integer(Str, 2),
    Top   = 1 bsl (Bits - 1),
    case Value band Top of
        0 -> Value;                              % positive
        _ -> Value - (1 bsl Bits)                % negative (subtract 2^Bits)
    end.

ensure_string(S) when is_list(S) -> S;
ensure_string(B) when is_binary(B) -> binary_to_list(B).
